src/manager/topdown.rs

```diff
@@ -0,0 +1,247 @@
+                            if curr_epoch >= submission_epoch {
+                                submit_topdown_checkpoint(
+                                    submission_epoch,
+                                    parent_tip_set,
+                                    child_tip_set,
+                                    account,
+                                    child.id.clone(),
+                                    &child_client,
+                                    &parent_client,
+                                )
+                                .await?;
+                            }
+                        }
+                    }
+                }
+            }
+
+            // Sleep for an appropriate amount of time before checking the chain head again or return
+            // if a stop notification is received.
+            if !wait_next_iteration(&stop_notify, CHAIN_HEAD_REQUEST_PERIOD).await? {
+                return Ok(());
+            }
+        }
+    };
+
+    result.context(format!(
+        "error in manage_topdown_checkpoints() for subnet pair ({}, {})",
+        parent.id, child.id
+    ))
+}
+
+// Prototype function for submitting topdown messages. This function is supposed to be called each
+// Nth epoch of a parent subnet. It reads the topdown messages from the parent subnet and submits
+// them to the child subnet.
+async fn submit_topdown_checkpoint<T: JsonRpcClient + Send + Sync>(
+    submission_epoch: ChainEpoch,
+    curr_parent_tip_set: Cid,
+    curr_child_tip_set: Cid,
+    account: &Address,
+    child_subnet: SubnetID,
+    child_client: &LotusJsonRPCClient<T>,
+    parent_client: &LotusJsonRPCClient<T>,
+) -> Result<()> {
+    log::debug!("Submitting topdown checkpoint for account {}", account);
+    // First, we read from the child subnet the nonce of the last topdown message executed
+    // after the last executed checkpoint. We
+    // increment the result by one to obtain the nonce of the first topdown message we want to
+    // submit to the child subnet.
+    let state = child_client
+        .ipc_read_gateway_state(curr_child_tip_set)
+        .await?;
+    let nonce = state.applied_topdown_nonce + 1;
+
+    // Then, we get the top-down messages from the latest nonce at the specific submission epoch.
+    // This ensures that all validators will provide deterministically the top-down messages
+    // to be included in all checkpoints.
+    // FIXME: We probably shouldn't use the default one here and use the one
+    // deployed in the subnet
+    let gateway_addr = Address::from_str(GATEWAY_ACTOR_ADDRESS)?;
+    let submission_tip_set = parent_client
+        .get_tipset_by_height(submission_epoch, curr_parent_tip_set)
+        .await?;
+    let cid_map = submission_tip_set.cids.first().unwrap().clone();
+    let submission_tip_set = Cid::try_from(cid_map)?;
+    let top_down_msgs = parent_client
+        .ipc_get_topdown_msgs(&child_subnet, gateway_addr, submission_tip_set, nonce)
```

Comment:
do we need some kind of pagination support to limit the return size? If there are a lot of top down msgs, it would blow up the gas?

Reply:
That's a great point. I will also track this one in an issue. I think this is a problem for the future, but we will definitely have to use some kind of logic to prevent from running out of gas inside a block. The thing is that we need this to be deterministic so all checkpoints include the same messages, if not this can end up in checkpoints never being executed.